File: src/main/java/com/baulsupp/oksocial/completion/UrlCompleter.java
package com.baulsupp.oksocial.completion;

import com.baulsupp.oksocial.authenticator.AuthInterceptor;
import com.baulsupp.oksocial.credentials.CredentialsStore;
import com.google.common.collect.Lists;
import java.io.IOException;
import java.time.Clock;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;
import java.util.logging.Level;
import java.util.logging.Logger;
import okhttp3.HttpUrl;
import okhttp3.OkHttpClient;

import static java.lang.Math.min;
import static java.util.regex.Pattern.quote;

public class UrlCompleter {
  private static Logger logger = Logger.getLogger(UrlCompleter.class.getName());

  private Iterable<AuthInterceptor<?>> services;
  private OkHttpClient client;
  private CredentialsStore credentialsStore;
  private CompletionVariableCache completionVariableCache;
  private Clock clock = Clock.systemDefaultZone();

  public UrlCompleter(List<AuthInterceptor<?>> services, OkHttpClient client,
      CredentialsStore credentialsStore, CompletionVariableCache completionVariableCache) {
    this.services = services;
    this.client = client;
    this.credentialsStore = credentialsStore;
    this.completionVariableCache = completionVariableCache;
  }

  public UrlList urlList(String prefix) throws IOException {

    Optional<HttpUrl> fullUrl = parseUrl(prefix);

    if (fullUrl.isPresent()) {
      HttpUrl u = fullUrl.get();

      for (AuthInterceptor<?> a : services) {
        if (a.supportsUrl(u)) {
          try {
            return a.apiCompleter(prefix, client, credentialsStore, completionVariableCache)
                .siteUrls(u)
                .get();
          } catch (InterruptedException e) {
            logger.log(Level.FINE, "interrupted", e);
          } catch (ExecutionException e) {
            logger.log(Level.WARNING, "completion failed", e.getCause());
          }
        }
      }

      // won't match anything
      return new UrlList("FAILED", Lists.newArrayList());
    } else {
      List<Future<UrlList>> futures = Lists.newArrayList();

      for (AuthInterceptor<?> a : services) {
        futures.add(
            a.apiCompleter("", client, credentialsStore, completionVariableCache).prefixUrls());
      }

      return futuresToList(prefix, futures);
    }
  }

  private UrlList futuresToList(String prefix, List<Future<UrlList>> futures) {
    long to = clock.millis() + 2000;

    List<String> results = Lists.newArrayList();

    for (Future<UrlList> f : futures) {
      try {
        UrlList result = f.get(to - clock.millis(), TimeUnit.MILLISECONDS);

        results.addAll(result.getUrls(prefix));
      } catch (ExecutionException e) {
        logger.log(Level.WARNING, "failure during url completion", e.getCause());
      } catch (InterruptedException | TimeoutException e) {
        logger.log(Level.FINE, "timeout during url completion", e);
      }
    }

    return new UrlList(quote(prefix), results);
  }

  private Optional<HttpUrl> parseUrl(String prefix) {
    if (isSingleApi(prefix)) {
      return Optional.ofNullable(HttpUrl.parse(prefix));
    } else {
      return Optional.empty();
    }
  }

  private boolean isSingleApi(String prefix) {
    return prefix.matches("https://[^/]+/.*");
  }

  public static boolean isPossibleAddress(String urlCompletion) {
    return urlCompletion.startsWith("https://".substring(0, min(urlCompletion.length(), 8)));
  }
}

File: src/main/kotlin/com/baulsupp/oksocial/completion/UrlCompleter.kt
package com.baulsupp.oksocial.completion

import com.baulsupp.oksocial.Main
import com.baulsupp.oksocial.Token
import com.baulsupp.oksocial.kotlin.client
import com.baulsupp.oksocial.util.ClientException
import kotlinx.coroutines.experimental.CancellationException
import kotlinx.coroutines.experimental.CommonPool
import kotlinx.coroutines.experimental.async
import kotlinx.coroutines.experimental.withTimeout
import okhttp3.HttpUrl
import java.lang.Math.min
import java.util.concurrent.ExecutionException
import java.util.concurrent.TimeUnit
import java.util.logging.Level
import java.util.logging.Logger

class UrlCompleter(val main: Main) : ArgumentCompleter {
  override suspend fun urlList(prefix: String, tokenSet: Token): UrlList {
    val fullUrl = parseUrl(prefix)

    return if (fullUrl != null) {
      pathCompletion(fullUrl, prefix, tokenSet)
    } else {
      hostCompletion(prefix, tokenSet)
    }
  }

  private suspend fun pathCompletion(fullUrl: HttpUrl, prefix: String, tokenSet: Token): UrlList {
    return (main.authenticatingInterceptor.services
      .firstOrNull { it.supportsUrl(fullUrl) }
      ?.apiCompleter(prefix, client, main.credentialsStore, main.completionVariableCache, tokenSet)
      ?.siteUrls(fullUrl, tokenSet)
      ?: UrlList(UrlList.Match.EXACT, listOf()))
  }

  private suspend fun UrlCompleter.hostCompletion(prefix: String, tokenSet: Token): UrlList {
    val futures = main.authenticatingInterceptor.services.map {
      async(CommonPool) {
        withTimeout(2, TimeUnit.SECONDS) {
          it.apiCompleter("", client, main.credentialsStore, main.completionVariableCache, tokenSet).prefixUrls()
        }
      }
    }

    val results = mutableListOf<String>()
    for (f in futures) {
      try {
        results.addAll(f.await().getUrls(""))
      } catch (e: ClientException) {
        logger.log(Level.WARNING, "http error during url completion", e)
      } catch (e: CancellationException) {
        logger.log(Level.WARNING, "failure during url completion", e.cause)
      } catch (e: ExecutionException) {
        logger.log(Level.WARNING, "failure during url completion", e.cause)
      }
    }
    return UrlList(UrlList.Match.HOSTS, results)
  }

  private fun parseUrl(prefix: String): HttpUrl? = if (isSingleApi(prefix)) {
    HttpUrl.parse(prefix)
  } else {
    null
  }

  private fun isSingleApi(prefix: String): Boolean = prefix.matches("https://[^/]+/.*".toRegex())

  companion object {
    private val logger = Logger.getLogger(UrlCompleter::class.java.name)

    fun isPossibleAddress(urlCompletion: String): Boolean {
      return urlCompletion.startsWith("https://".substring(0, min(urlCompletion.length, 8)))
    }
  }
}
